pint/facets/numpy/quantity.py

```diff
@@ -245,7 +245,12 @@ def __getitem__(self, key):
 
     def __setitem__(self, key, value):
         try:
-            if np.ma.is_masked(value) or math.isnan(value):
+            # If we're dealing with a masked single value or a nan, set it
+            if (
+                isinstance(self._magnitude, np.ma.MaskedArray)
+                and np.ma.is_masked(value)
+                and getattr(value, "size", 0) == 1
+            ) or math.isnan(value):
```

Comment:
Have you considered using
```suggestion
            if (value is np.ma.masked) or math.isnan(value):
```
Then your requirement that `a[1] = np.ma.masked` works should be fulfilled, but you may avoid the other problems.

Reply:
I did contemplate that and ended up looking back when I contributed the original change. That specifically mentioned masked values. I think I can live with this version too (emphasis on *think*), but I'd want to hear thoughts from @hgrecco or any of the other core committers.

And for absolutely clarity: changing to that won't change the fact that setting using a masked array (with masked values) can completely subvert the units.